contracts/token-distribution/SaleRounds.sol

```diff
@@ -204,115 +159,186 @@ contract SaleRounds is TokenDistribution, GameOwner, ERC20 {
+        address exchangesWalletAddress = walletAddresses[0];
+        address playAndEarnWalletAddress = walletAddresses[1];
+        address socialWalletAddress = walletAddresses[2];
+        address teamWalletAddress = walletAddresses[3];
+        address treasuryWalletAddress = walletAddresses[4];
+        address advisorsWalletAddress = walletAddresses[5];
 
-        return reservedBalances[roundType][_to];
+        //ALLOCATIONS WITH WALLET CONSTANT
+        require(playAndEarnWalletAddress != address(0), "Play and earn wallet address is 0x0");
+        reserveTokensInternal(RoundType.PLAYANDEARN, playAndEarnWalletAddress, playAndEarnDistribution.supply);
+
+        require(socialWalletAddress != address(0), "Social wallet address is 0x0");
+        reserveTokensInternal(RoundType.SOCIAL, socialWalletAddress, socialDistribution.supply);
+
+        require(teamWalletAddress != address(0), "Team wallet address is 0x0");
+        reserveTokensInternal(RoundType.TEAM, teamWalletAddress, teamDistribution.supply);
+
+        require(treasuryWalletAddress != address(0), "Treasury wallet address is 0x0");
+        reserveTokensInternal(RoundType.TREASURY, treasuryWalletAddress, treasuryDistribution.supply);
+
+        require(advisorsWalletAddress != address(0), "Advisors wallet address is 0x0");
+        reserveTokensInternal(RoundType.ADVISOR, advisorsWalletAddress, advisorsDistribution.supply);
+
+        require(exchangesWalletAddress != address(0), "Exchanges wallet address is 0x0");
+
+        // Initial minting of 40% of total supply for exchanges distribution
+        uint256 initialExchangesSupply = 60_000_000;
+        _mint(exchangesWalletAddress, initialExchangesSupply);
+        roundDistribution[RoundType.EXCHANGES].totalRemaining -= initialExchangesSupply;
+        claimedBalances[RoundType.EXCHANGES][exchangesWalletAddress] += initialExchangesSupply;
+
+        // Reserving rest of the 60% of the supply for exchanges distribution
+        reserveTokensInternal(RoundType.EXCHANGES, exchangesWalletAddress, exchangesDistribution.supply - initialExchangesSupply);
     }
 
-    function setCliffTime(string calldata _roundType, uint256 _amount) public onlyGameOwner {
-        RoundType roundType =  getRoundTypeByKey(_roundType);
+    function calculateCliffTimeDiff(ClaimInfo memory claimInfo) private view returns(uint) {
+        //How many seconds since the cliff? (negative if before cliff)
+        (, uint timeDiff) = (block.timestamp - claimInfo.startTime).trySub(claimInfo.cliff);
+        return timeDiff;
+    }
 
-        roundDistribution[roundType].cliff = _amount;
+    function calculateVestingForUserPerSecond(ClaimInfo memory claimInfo) private pure returns(uint) {
+        //We divide the balance by the number of seconds in the entire vesting period (vesting unit is seconds!).
+        (, uint vestingForUserPerSecond) = claimInfo.balance.tryDiv(claimInfo.vesting);
```

Comment:
Regarding my question above, there may be a bug here - what if someone has fewer tokens reserved than there are seconds in their vesting period? Then the code referenced above might insta-claim all of them.

Reply:
I remember what the issue was, with 18 decimals this basically cannot happen. :)